File: system-test/java/src/main/java/com/optivem/shop/testkit/core/usecase/external/tax/TaxDsl.java
package com.optivem.shop.testkit.core.usecase.external.tax;

import com.optivem.shop.testkit.common.Closer;
import com.optivem.shop.testkit.core.shared.UseCaseContext;
import com.optivem.shop.testkit.core.usecase.external.tax.usecases.GetTaxRate;
import com.optivem.shop.testkit.core.usecase.external.tax.usecases.GoToTax;
import com.optivem.shop.testkit.core.usecase.external.tax.usecases.ReturnsTaxRate;
import com.optivem.shop.testkit.driver.port.external.tax.TaxDriver;

public class TaxDsl implements AutoCloseable {
    private final TaxDriver driver;
    private final UseCaseContext context;

    public TaxDsl(TaxDriver driver, UseCaseContext context) {
        this.driver = driver;
        this.context = context;
    }

    @Override
    public void close() {
        Closer.close(driver);
    }

    public GoToTax goToTax() {
        return new GoToTax(driver, context);
    }

    public ReturnsTaxRate returnsTaxRate() {
        return new ReturnsTaxRate(driver, context);
    }

    public GetTaxRate getTaxRate() {
        return new GetTaxRate(driver, context);
    }
}

File: system-test/typescript/src/testkit/dsl/core/usecase/external/tax/TaxDsl.ts
import type { TaxDriver } from '../../../../../driver/port/external/tax/tax-driver.js';
import type { UseCaseContext } from '../../../use-case-context.js';
import { GoToTax } from './usecases/GoToTax.js';
import { GetTaxRate } from './usecases/GetTaxRate.js';
import { ReturnsTaxRate } from './usecases/ReturnsTaxRate.js';

export class TaxDsl {
  constructor(
    private readonly driver: TaxDriver,
    private readonly context: UseCaseContext,
  ) {}

  goToTax(): GoToTax {
    return new GoToTax(this.driver, this.context);
  }

  getTaxRate(): GetTaxRate {
    return new GetTaxRate(this.driver, this.context);
  }

  returnsTaxRate(): ReturnsTaxRate {
    return new ReturnsTaxRate(this.driver, this.context);
  }

  async close(): Promise<void> {
    await this.driver.close();
  }
}
